### PR TITLE
fix: disk cache deduped get_ranges

### LINKED ISSUE #1215: Error getting data from disk cache in concurrent situation
### Describe this problem
see titile

```
---- disk_cache::test::test_disk_cache_multi_thread_fetch_same_block stdout ----
thread 'disk_cache::test::test_disk_cache_multi_thread_fetch_same_block' panicked at components/object_store/src/disk_cache.rs:1107:13:
assertion `left == right` failed
  left: b"i j k l m n o p q r s t u v w x y za b c d e f g\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0 p q r s t u v w x y"
 right: b"i j k l m n o p q r s t u v w x y za b c d e f g h i j k l m n o p q r s t u v w x y"
stack backtrace:
   0: rust_begin_unwind
             at /rustc/8550f15e148407159af401e02b1d9259762b3496/library/std/src/panicking.rs:617:5
   1: core::panicking::panic_fmt
             at /rustc/8550f15e148407159af401e02b1d9259762b3496/library/core/src/panicking.rs:67:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/8550f15e148407159af401e02b1d9259762b3496/library/core/src/panicking.rs:248:5
   4: object_store::disk_cache::test::test_disk_cache_multi_thread_fetch_same_block::{{closure}}
             at ./src/disk_cache.rs:1107:13
   5: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/8550f15e148407159af401e02b1d9259762b3496/library/core/src/future/future.rs:125:9
   6: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/8550f15e148407159af401e02b1d9259762b3496/library/core/src/future/future.rs:125:9
   7: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/scheduler/current_thread.rs:651:57
   8: tokio::runtime::coop::with_budget
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/coop.rs:107:5
   9: tokio::runtime::coop::budget
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/coop.rs:73:5
  10: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/scheduler/current_thread.rs:651:25
  11: tokio::runtime::scheduler::current_thread::Context::enter
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/scheduler/current_thread.rs:410:19
  12: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/scheduler/current_thread.rs:650:36
  13: tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/scheduler/current_thread.rs:729:68
  14: tokio::runtime::context::scoped::Scoped<T>::set
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/context/scoped.rs:40:9
  15: tokio::runtime::context::set_scheduler::{{closure}}
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/context.rs:176:26
  16: std::thread::local::LocalKey<T>::try_with
             at /rustc/8550f15e148407159af401e02b1d9259762b3496/library/std/src/thread/local.rs:270:16
  17: std::thread::local::LocalKey<T>::with
             at /rustc/8550f15e148407159af401e02b1d9259762b3496/library/std/src/thread/local.rs:246:9
  18: tokio::runtime::context::set_scheduler
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/context.rs:176:9
  19: tokio::runtime::scheduler::current_thread::CoreGuard::enter
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/scheduler/current_thread.rs:729:27
  20: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/scheduler/current_thread.rs:638:19
  21: tokio::runtime::scheduler::current_thread::CurrentThread::block_on::{{closure}}
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/scheduler/current_thread.rs:175:28
  22: tokio::runtime::context::runtime::enter_runtime
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/context/runtime.rs:65:16
  23: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/scheduler/current_thread.rs:167:9
  24: tokio::runtime::runtime::Runtime::block_on
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/runtime.rs:311:47
  25: object_store::disk_cache::test::test_disk_cache_multi_thread_fetch_same_block
             at ./src/disk_cache.rs:1106:9
  26: object_store::disk_cache::test::test_disk_cache_multi_thread_fetch_same_block::{{closure}}
             at ./src/disk_cache.rs:[1066](https://github.com/CeresDB/ceresdb/actions/runs/6183763154/job/16786118811?pr=1214#step:8:1067):62
  27: core::ops::function::FnOnce::call_once
             at /rustc/8550f15e148407159af401e02b1d9259762b3496/library/core/src/ops/function.rs:250:5
  28: core::ops::function::FnOnce::call_once
             at /rustc/8550f15e148407159af401e02b1d9259762b3496/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
https://github.com/CeresDB/ceresdb/actions/runs/6183763154/job/16786118811?pr=1214

### Server version

a7f09dfd509f5ba770ab35301125a1a32ab5b66d

### Steps to reproduce

CI

### Expected behavior

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by jiacai2050**
## Rationale
Close #1215

## Detailed Changes
- Add MemoryStore to help test
- When persist file, first write to tmp file, then rename to dest file.

## Test Plan
CI itself

### CODE REVIEW

File: components/object_store/src/disk_cache.rs [hunk 1]
```diff
@@ -1103,6 +1103,7 @@ mod test {
         }
 
         let actual = futures::future::join_all(tasks).await;
+        println!("get_counts, {}", store.inner.underlying_store);
```
Comment (tanruixiang) [inline components/object_store/src/disk_cache.rs:1106]:
```suggestion
```

File: components/object_store/src/disk_cache.rs [hunk 2]
```diff
@@ -357,18 +357,36 @@ impl DiskCache {
...
+        // When write payload to file, the cache lock is released, so when one thread is
+        // reading, another thread may update it, so we write to tmp file first,
+        // then rename to expected filename to avoid other threads see partial
+        // content.
+        let tmp_filename = format!("{filename}.tmp");
+        let tmp_filepath = std::path::Path::new(&self.root_dir)
+            .join(&tmp_filename)
             .into_os_string()
             .into_string()
             .unwrap();
```
Comment (ShiKaiWi) [inline components/object_store/src/disk_cache.rs:369]: Shall we introduce a specific implementation for such `tempfile-for-renaming`? At least, we should ensure the unused temp file to be cleaned if any error occurs after it is created.

File: components/object_store/src/lib.rs [hunk 3]
```diff
@@ -30,5 +30,7 @@ pub mod multipart;
 pub mod obkv;
 pub mod prefix;
 pub mod s3;
+#[cfg(test)]
```
Comment (ShiKaiWi) [inline components/object_store/src/lib.rs:33]: Is this module used by other crate?

File: components/object_store/src/disk_cache.rs [hunk 4]
```diff
@@ -167,36 +167,63 @@ impl Manifest {
...
+    // reading, another thread may update it, so we write to tmp file first,
+    // then rename to expected filename to avoid other threads see partial
+    // content.
+    async fn write_and_flush(self, bytes: Bytes) -> Result<()> {
+        let tmp_file = Self::tmp_file(&self.output);
+        let write_result = self.write_inner(&tmp_file, bytes).await;
+        if write_result.is_err() {
+            // we don't care this result.
+            _ = tokio::fs::remove_file(&tmp_file).await;
+        }
```
Comment (ShiKaiWi) [inline components/object_store/src/disk_cache.rs:222]: It's not a better way to remove tempfile. Maybe we should follow RAII style to trigger tempfile removal even if the whole scope (maybe it's tokio task) is dropped.

File: components/object_store/src/lib.rs (hunk 3 above)
Comment (jiacai2050) [reply components/object_store/src/lib.rs:33]: Only in test now.

File: components/object_store/src/disk_cache.rs (hunk 4 above)
Comment (jiacai2050) [reply components/object_store/src/disk_cache.rs:222]: We can't execute async fn in `drop` directly, and I don't want to spawn a runtime inside it.

**Review by ShiKaiWi — APPROVED** [review]

**Pull request merged**